packages/session-replay/src/BacktraceSessionRecorder.ts

```diff
@@ -0,0 +1,80 @@
+import { BacktraceAttachment } from '@backtrace/sdk-core';
+import { eventWithTime } from '@rrweb/types';
+import { record } from 'rrweb';
+import { BacktraceSessionRecorderOptions } from './options';
+
+function defaultIfNotFalse<T>(value: T | false, defaultValue?: T) {
+    if (value === false) {
+        return undefined;
+    }
+
+    return value ?? defaultValue;
+}
+
+export class BacktraceSessionRecorder implements BacktraceAttachment {
+    public readonly name = 'bt-session-replay-0';
+    public readonly type = 'dynamic';
+
+    private readonly _maxEventCount?: number;
+    private readonly _maxTime?: number;
+
+    private _previousEvents?: eventWithTime[] = [];
+    private _events?: eventWithTime[] = [];
+
+    private _stop?: () => void;
+
+    constructor(private readonly _options: BacktraceSessionRecorderOptions) {
+        this._events = [];
+        this._maxEventCount = defaultIfNotFalse(_options.maxEventCount, 100);
+        this._maxTime = defaultIfNotFalse(_options.maxTime, undefined);
+    }
+
+    public start() {
+        this._stop = record({
+            ...this._options.advancedOptions,
+            sampling: {
+                mousemove: this._options.sampling?.mousemove,
+                mouseInteraction: this._options.sampling?.mouseInteraction,
+                input: this._options.sampling?.input,
+                media: this._options.sampling?.media,
+                scroll: this._options.sampling?.scroll,
+                ...this._options.advancedOptions,
+            },
+            emit: (event, isCheckout) => this.onEmit(event, isCheckout),
+            checkoutEveryNth: this._maxEventCount && Math.ceil(this._maxEventCount / 2),
```

Comment:
Each event group has to be "checked out" by `rrweb`, i.e. made sure that it can function on it's own. If you have a maximum of 100 events, the checkout will be every 50 events to make sure that we do sent at least 50 events after reset and not 0. If we would set it to 100, we would be sending a maximum of 200 events if we didn't want to clear the first group. Not sure if I explained this clearly 😅

Reply:
So it's to give essentially two chances (50 events, 50 events) for a max of 100 in case the checkout of the first 50 fails, for example?

Reply:
It's goes more like this:
1. add first event, checkout - initialize empty `_events` array and add the event to it
2. add 49 events (so we have 50 events)
3. add 51st event, checkout - move `_events` to `_previousEvents`, reinitialize empty `_events` array and add the event to it
4. add 49 events (so we have 100 events)
5. add 101st event, checkout - move `_events` to `_previousEvents` (that drops the previous `_previousEvents`), reinitialize empty `_events` array and add the event to it
6. and so on...

This ensures that after adding 50 events, 50-100 events will be always available, and we do not exceed the 100 events limit.

Reply:
Ahh, got it. Thanks!